From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Mon, 26 Dec 2022 13:47:08 +0100
Subject: [PATCH] Fiddle branding changes
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

No license is given for images and license text files.
Images are copyright (©) FiddleMC and its contributors, all rights reserved.
License text files are copyright (©) their respective authors.
License for other files: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

This patch is based on the following patch:
"Gale branding changes"
By: Martijn Muijsers <martijnmuijsers@live.nl>
As part of: Gale (https://github.com/GaleMC/Gale)
Licensed (except for image files) under: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

* Gale description *

Gale - https://galemc.org

diff --git a/build.gradle.kts b/build.gradle.kts
index 4073145e01b34ec31ce16e5a75f4634a6c192e99..97d35399f7e598c2dc1ebf08448075969385d3a8 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -90,7 +90,7 @@ tasks.jar {
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Paper-$implementationVersion",
+            "Implementation-Version" to "git-Fiddle-$implementationVersion", // Fiddle - branding changes
             "Implementation-Vendor" to date, // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
diff --git a/license/FPL-1.0.txt b/license/FPL-1.0.txt
new file mode 100644
index 0000000000000000000000000000000000000000..5217f52bf08e479525ab4af9af20869da33c26ed
--- /dev/null
+++ b/license/FPL-1.0.txt
@@ -0,0 +1,86 @@
+Below is the Fiddle Public License Agreement 1.0 (FPL-1.0),
+which applies only to parts of Fiddle that are specifically and
+explicitly released under FPL-1.0.
+
+                FIDDLE PUBLIC LICENSE AGREEMENT
+                    Version 1.0, 29 May 2023
+
+ Copyright (C) 2023 Martijn Muijsers
+ Everyone is permitted to copy and distribute verbatim copies
+ of this license document, but changing it is not allowed.
+
+  For pre-existing (typically as part of the Minecraft, Bukkit,
+CraftBukkit, Spigot or Paper source code) files patched with Fiddle
+patches by the Paperweight patch system, any patch hunks that apply to
+those files are licensed under the GNU General Public License, version 3
+(GPL-3.0).
+
+  Similarly, any components (such as patch hunks, source files or
+included libraries) in a composite product (such as a repository of
+source code or a jar release file) that is released under this license
+(going forward, to be understood as the Fiddle Public License Agreement
+1.0) as a whole, and which are individually licensed under MIT, GPL-3.0
+or as part of an official Minecraft product (such as the Minecraft server
+source code, which is restrictively licensed as per Microsoft's terms),
+and which are not part of the modificiations made by Fiddle or any
+derivative thereof, are not subject to the further conditions of this
+license, but are instead licensed under their respective original
+licenses.
+
+  For files newly created with Fiddle patches (such as Java source code
+files to merge resource packs), any patch hunks that apply to those files
+are licensed subject to GPL-3.0 with the alterations that follow in this
+license.
+
+1. Share-alike
+
+  Where the GPL-3.0 license would require releasing something (whether
+pertaining to source code, binaries or otherwise) under a license that
+conforms to some particular requirements, such as GPL-3.0, the
+requirement instead becomes to release under this license.
+
+2. Paywalled content must pay-it-forward
+
+  You may not use Fiddle or a derivative thereof to provide a service to
+users, where some features are only available for payment, or more easily
+accessible after some payment than otherwise, except if you clearly
+credit Fiddle and the authors and contributors to the corresponding
+features and their assets, and where possible, pass a fair part (as would
+be considered fair by the community of the receiving parties, or in the
+form of a set amount or ratio as chosen by the author of the features) of
+the revenue on to those parties.
+  As an example, if you install a free pack that adds content, but only
+allow donators to use some parts of that content, then you must donate
+a fair part of the donations to that pack's authors and Fiddle.
+
+3. Prohibition on re-selling the server jar
+
+  Fiddle or derivatives thereof, including any resulting server jar, do
+not require having their source code shared just for running them as a
+Minecraft server, similar to GPL-3.0.
+  This means, that as a server owner, you may make additional custom
+modifications to the Fiddle source code, and use the resulting server jar
+as your Minecraft server.
+
+  However, distribution of Fiddle or derivatives thereof, including
+any resulting server jar, is not allowed to be conditional, whether
+financial or otherwise, to any extent.
+  For example, as a server host you are not allowed to place any
+conditions on server owners to use Fiddle or a derivative, compared to
+the corresponding upstream or an alternative (such as Paper or Fabric) or
+provide any cheaper or less conditional package at all that does not
+support Fiddle, but supports an upstream or alternative thereof.
+  Providing access to the functionality of Fiddle or derivatives thereof
+in any way, whether directly or indirectly, that is not directed at an
+end user Minecraft client, is considered distribution.
+  Additionally, any such distribution of a Fiddle derivative that is not
+directed at an end user Minecraft client requires the full source code of
+the additional modifications in the modified version (as defined by
+GPL-3.0) to be pre-emptively, publicly (specifically, as a public
+repository on GitHub) released under the Fiddle Public License Agreement
+1.0, and a reference to that release must be clearly displayed to all
+intermediate users (such as server owners) and end users (such as
+Minecraft clients).
+  For example, distributing a derivative of Fiddle that contains
+performance enhancements requires the full source code for the
+modifications to be pre-emptively, publicly released.
\ No newline at end of file
diff --git a/LICENCE.txt b/license/GPL-3.0.txt
similarity index 98%
rename from LICENCE.txt
rename to license/GPL-3.0.txt
index 94a9ed024d3859793618152ea559a168bbcbb5e2..2d5b975212128baa0b84f9ec723491f54940b188 100644
--- a/LICENCE.txt
+++ b/license/GPL-3.0.txt
@@ -1,7 +1,10 @@
+Below is the GNU General Public License, version 3 (GPL-3.0),
+which applies only to parts of Fiddle that are specifically and explicitly released under GPL-3.0.
+
                     GNU GENERAL PUBLIC LICENSE
                        Version 3, 29 June 2007
 
- Copyright (C) 2007 Free Software Foundation, Inc. <http://fsf.org/>
+ Copyright (C) 2007 Free Software Foundation, Inc. <https://fsf.org/>
  Everyone is permitted to copy and distribute verbatim copies
  of this license document, but changing it is not allowed.
 
@@ -645,7 +648,7 @@ the "copyright" line and a pointer to where the full notice is found.
     GNU General Public License for more details.
 
     You should have received a copy of the GNU General Public License
-    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 Also add information on how to contact you by electronic and paper mail.
 
@@ -664,11 +667,11 @@ might be different; for a GUI interface, you would use an "about box".
   You should also get your employer (if you work as a programmer) or school,
 if any, to sign a "copyright disclaimer" for the program, if necessary.
 For more information on this, and how to apply and follow the GNU GPL, see
-<http://www.gnu.org/licenses/>.
+<https://www.gnu.org/licenses/>.
 
   The GNU General Public License does not permit incorporating your program
 into proprietary programs.  If your program is a subroutine library, you
 may consider it more useful to permit linking proprietary applications with
 the library.  If this is what you want to do, use the GNU Lesser General
 Public License instead of this License.  But first, please read
-<http://www.gnu.org/philosophy/why-not-lgpl.html>.
+<https://www.gnu.org/licenses/why-not-lgpl.html>.
diff --git a/LGPL.txt b/license/LGPL-3.0.txt
similarity index 97%
rename from LGPL.txt
rename to license/LGPL-3.0.txt
index 65c5ca88a67c30becee01c5a8816d964b03862f9..0d0f2ad656e123e3c9e289706363f93b7c9edda3 100644
--- a/LGPL.txt
+++ b/license/LGPL-3.0.txt
@@ -1,3 +1,6 @@
+Below is the GNU Lesser General Public License, version 3 (LGPL-3.0),
+which applies only to parts of Fiddle that are specifically and explicitly released under LGPL-3.0.
+
                    GNU LESSER GENERAL PUBLIC LICENSE
                        Version 3, 29 June 2007
 
diff --git a/license/MIT.txt b/license/MIT.txt
new file mode 100644
index 0000000000000000000000000000000000000000..4cd1c5d76394f8f7f99005563a75b4275decae41
--- /dev/null
+++ b/license/MIT.txt
@@ -0,0 +1,24 @@
+Below is the MIT License (MIT),
+which applies only to parts of Fiddle that are specifically and explicitly released under MIT.
+
+MIT License
+
+Copyright (c) 2022-2023 Respective authors of parts of Fiddle or FiddleMC software, or software included therein
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
\ No newline at end of file
diff --git a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
index 9d687da5bdf398bb3f6c84cdf1249a7213d09f2e..0393a61ef229868568fe8127e71d9c442d835a58 100644
--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -1,85 +1,40 @@
 package com.destroystokyo.paper;
 
-import com.destroystokyo.paper.util.VersionFetcher;
 import com.google.common.base.Charsets;
 import com.google.common.io.Resources;
-import com.google.gson.*;
-import net.kyori.adventure.text.Component;
-import net.kyori.adventure.text.event.ClickEvent;
-import net.kyori.adventure.text.format.NamedTextColor;
-import net.kyori.adventure.text.format.TextDecoration;
-import net.kyori.adventure.text.TextComponent;
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+import org.fiddlemc.fiddle.version.AbstractPaperVersionFetcher;
 
-import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
-import java.io.*;
-import java.net.HttpURLConnection;
+import java.io.BufferedReader;
+import java.io.IOException;
 import java.net.URL;
 import java.util.stream.StreamSupport;
 
-public class PaperVersionFetcher implements VersionFetcher {
-    private static final java.util.regex.Pattern VER_PATTERN = java.util.regex.Pattern.compile("^([0-9\\.]*)\\-.*R"); // R is an anchor, will always give '-R' at end
-    private static final String GITHUB_BRANCH_NAME = "master";
-    private static final String DOWNLOAD_PAGE = "https://papermc.io/downloads/paper";
-    private static @Nullable String mcVer;
+// Fiddle start - branding changes - version fetcher
+/**
+ * The original version fetcher for Paper. Most of the original content of this class has been moved to
+ * {@link AbstractPaperVersionFetcher}.
+ */
+public class PaperVersionFetcher extends AbstractPaperVersionFetcher {
 
-    @Override
-    public long getCacheTime() {
-        return 720000;
+    public PaperVersionFetcher() {
+        super("master", "https://papermc.io/downloads/paper", "Paper", "PaperMC", "PaperMC", "Paper");
     }
 
-    @Nonnull
     @Override
-    public Component getVersionMessage(@Nonnull String serverVersion) {
-        String[] parts = serverVersion.substring("git-Paper-".length()).split("[-\\s]");
-        final Component updateMessage = getUpdateStatusMessage("PaperMC/Paper", GITHUB_BRANCH_NAME, parts[0]);
-        final Component history = getHistory();
-
-        return history != null ? TextComponent.ofChildren(updateMessage, Component.newline(), history) : updateMessage;
-    }
-
-    private static @Nullable String getMinecraftVersion() {
-        if (mcVer == null) {
-            java.util.regex.Matcher matcher = VER_PATTERN.matcher(org.bukkit.Bukkit.getBukkitVersion());
-            if (matcher.find()) {
-                String result = matcher.group();
-                mcVer = result.substring(0, result.length() - 2); // strip 'R' anchor and trailing '-'
-            } else {
-                org.bukkit.Bukkit.getLogger().warning("Unable to match version to pattern! Report to PaperMC!");
-                org.bukkit.Bukkit.getLogger().warning("Pattern: " + VER_PATTERN.toString());
-                org.bukkit.Bukkit.getLogger().warning("Version: " + org.bukkit.Bukkit.getBukkitVersion());
-            }
-        }
-
-        return mcVer;
+    protected boolean canFetchDistanceFromSiteApi() {
+        return true;
     }
 
-    private static Component getUpdateStatusMessage(@Nonnull String repo, @Nonnull String branch, @Nonnull String versionInfo) {
-        int distance;
-        try {
-            int jenkinsBuild = Integer.parseInt(versionInfo);
-            distance = fetchDistanceFromSiteApi(jenkinsBuild, getMinecraftVersion());
-        } catch (NumberFormatException ignored) {
-            versionInfo = versionInfo.replace("\"", "");
-            distance = fetchDistanceFromGitHub(repo, branch, versionInfo);
-        }
-
-        switch (distance) {
-            case -1:
-                return Component.text("Error obtaining version information", NamedTextColor.YELLOW);
-            case 0:
-                return Component.text("You are running the latest version", NamedTextColor.GREEN);
-            case -2:
-                return Component.text("Unknown version", NamedTextColor.YELLOW);
-            default:
-                return Component.text("You are " + distance + " version(s) behind", NamedTextColor.YELLOW)
-                        .append(Component.newline())
-                        .append(Component.text("Download the new version at: ")
-                                .append(Component.text(DOWNLOAD_PAGE, NamedTextColor.GOLD)
-                                        .hoverEvent(Component.text("Click to open", NamedTextColor.WHITE))
-                                        .clickEvent(ClickEvent.openUrl(DOWNLOAD_PAGE))));
-        }
+    @Override
+    protected int fetchDistanceFromSiteApi(int jenkinsBuild) {
+        return fetchDistanceFromSiteApi(jenkinsBuild, this.getMinecraftVersion());
     }
+    // Fiddle end - branding changes - version fetcher
 
     private static int fetchDistanceFromSiteApi(int jenkinsBuild, @Nullable String siteApiVersion) {
         if (siteApiVersion == null) { return -1; }
@@ -105,45 +60,4 @@ public class PaperVersionFetcher implements VersionFetcher {
         }
     }
 
-    // Contributed by Techcable <Techcable@outlook.com> in GH-65
-    private static int fetchDistanceFromGitHub(@Nonnull String repo, @Nonnull String branch, @Nonnull String hash) {
-        try {
-            HttpURLConnection connection = (HttpURLConnection) new URL("https://api.github.com/repos/" + repo + "/compare/" + branch + "..." + hash).openConnection();
-            connection.connect();
-            if (connection.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) return -2; // Unknown commit
-            try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), Charsets.UTF_8))) {
-                JsonObject obj = new Gson().fromJson(reader, JsonObject.class);
-                String status = obj.get("status").getAsString();
-                switch (status) {
-                    case "identical":
-                        return 0;
-                    case "behind":
-                        return obj.get("behind_by").getAsInt();
-                    default:
-                        return -1;
-                }
-            } catch (JsonSyntaxException | NumberFormatException e) {
-                e.printStackTrace();
-                return -1;
-            }
-        } catch (IOException e) {
-            e.printStackTrace();
-            return -1;
-        }
-    }
-
-    @Nullable
-    private Component getHistory() {
-        final VersionHistoryManager.VersionData data = VersionHistoryManager.INSTANCE.getVersionData();
-        if (data == null) {
-            return null;
-        }
-
-        final String oldVersion = data.getOldVersion();
-        if (oldVersion == null) {
-            return null;
-        }
-
-        return Component.text("Previous version: " + oldVersion, NamedTextColor.GRAY, TextDecoration.ITALIC);
-    }
 }
diff --git a/src/main/java/com/destroystokyo/paper/console/PaperConsole.java b/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
index c5d5648f4ca603ef2b1df723b58f9caf4dd3c722..374acd34f435ec98b9f9b5851a316f2aaad97b95 100644
--- a/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
+++ b/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
@@ -17,7 +17,7 @@ public final class PaperConsole extends SimpleTerminalConsole {
     @Override
     protected LineReader buildReader(LineReaderBuilder builder) {
         builder
-                .appName("Paper")
+                .appName("Fiddle") // Fiddle - branding changes
                 .variable(LineReader.HISTORY_FILE, java.nio.file.Paths.get(".console_history"))
                 .completer(new ConsoleCommandCompleter(this.server))
                 .option(LineReader.Option.COMPLETE_IN_WORD, true);
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 3238cbcba567b1242c77e41f6b6f19a8d157fb4e..2783399f78d2df48328cc4d235c6ded065679227 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -921,7 +921,13 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         shutdownThread = Thread.currentThread();
         org.spigotmc.WatchdogThread.doStop(); // Paper
         if (!isSameThread()) {
-            MinecraftServer.LOGGER.info("Stopping main thread (Ignore any thread death message you see! - DO NOT REPORT THREAD DEATH TO PAPER)");
+            // Fiddle start - branding changes
+            /*
+            We do not want people to report thread issues to Paper,
+            but we do want people to report thread issues to Fiddle.
+             */
+            MinecraftServer.LOGGER.info("Stopping main thread (Ignore any thread death message you see! - DO NOT REPORT THREAD DEATH TO PAPER - If you think this is a Fiddle bug, please report it at https://github.com/FiddleMC/Fiddle/issues )");
+            // Fiddle end - branding changes
             while (this.getRunningThread().isAlive()) {
                 this.getRunningThread().stop();
                 try {
@@ -1693,7 +1699,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return "Paper"; // Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
+        return "Fiddle"; // Fiddle - branding changes - Fiddle > // Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
     }
 
     public SystemReport fillSystemReport(SystemReport details) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index a05c1bc8874ef5e380544a6a344c848e37da49c4..d5af5f7f7af4a2a94b15395962472be8952ef63c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -264,7 +264,7 @@ import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
 
 public final class CraftServer implements Server {
-    private final String serverName = "Paper"; // Paper
+    private final String serverName = "Fiddle"; // Paper // Fiddle - branding changes
     private final String serverVersion;
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
index d7ce4971d9271dbeff4adb9d852e4e7bdf60bf03..25a08e709afa849ac2f09b6f1a4df2396a561316 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
@@ -501,7 +501,7 @@ public class CraftScheduler implements BukkitScheduler {
                 this.parsePending();
             } else {
                 // this.debugTail = this.debugTail.setNext(new CraftAsyncDebugger(currentTick + CraftScheduler.RECENT_TICKS, task.getOwner(), task.getTaskClass())); // Paper
-                task.getOwner().getLogger().log(Level.SEVERE, "Unexpected Async Task in the Sync Scheduler. Report this to Paper"); // Paper
+                task.getOwner().getLogger().log(Level.SEVERE, "Unexpected Async Task in the Sync Scheduler. Report this to Fiddle"); // Paper // Fiddle - branding changes
                 // We don't need to parse pending
                 // (async tasks must live with race-conditions if they attempt to cancel between these few lines of code)
             }
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 83cef5ec27c31f133a23cd27349f722799c786ea..90a3f82a2491379318db6382ed8b8f80524dbc0a 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -470,7 +470,7 @@ public final class CraftMagicNumbers implements UnsafeValues {
 
     @Override
     public com.destroystokyo.paper.util.VersionFetcher getVersionFetcher() {
-        return new com.destroystokyo.paper.PaperVersionFetcher();
+        return new org.fiddlemc.fiddle.version.FiddleVersionFetcher(); // Fiddle - branding changes - version fetcher
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index 774556a62eb240da42e84db4502e2ed43495be17..900bbc05e6b9dfc2118893989e17c5658635abba 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -11,7 +11,7 @@ public final class Versioning {
     public static String getBukkitVersion() {
         String result = "Unknown-Version";
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/io.papermc.paper/paper-api/pom.properties");
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/org.fiddlemc.fiddle/fiddle-api/pom.properties"); // Fiddle - branding changes
         Properties properties = new Properties();
 
         if (stream != null) {
diff --git a/src/main/java/org/fiddlemc/fiddle/version/AbstractPaperVersionFetcher.java b/src/main/java/org/fiddlemc/fiddle/version/AbstractPaperVersionFetcher.java
new file mode 100644
index 0000000000000000000000000000000000000000..4a942eb75d5ccf9633b66a51e6208732c5da9b43
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/version/AbstractPaperVersionFetcher.java
@@ -0,0 +1,179 @@
+// Fiddle - branding changes - version fetcher
+
+package org.fiddlemc.fiddle.version;
+
+import com.destroystokyo.paper.PaperVersionFetcher;
+import com.destroystokyo.paper.VersionHistoryManager;
+import com.destroystokyo.paper.util.VersionFetcher;
+import com.google.common.base.Charsets;
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.event.ClickEvent;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextDecoration;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+/**
+ * An abstract version fetcher, derived from {@link PaperVersionFetcher}.
+ * This class was then made to be a superclass of both {@link PaperVersionFetcher}
+ * and {@link FiddleVersionFetcher}.
+ * <br>
+ * Changes to {@link PaperVersionFetcher} are indicated by Fiddle marker comments.
+ */
+public abstract class AbstractPaperVersionFetcher implements VersionFetcher {
+    private static final java.util.regex.Pattern VER_PATTERN = java.util.regex.Pattern.compile("^([0-9\\.]*)\\-.*R"); // R is an anchor, will always give '-R' at end
+    private static @Nullable String mcVer;
+
+    // Fiddle start - branding changes - version fetcher
+
+    protected final String gitHubBranchName;
+    protected final String downloadPage;
+    protected final String projectDisplayName;
+    protected final String organizationDisplayName;
+    protected final String gitHubOrganizationName;
+    protected final String gitHubRepoName;
+
+    protected AbstractPaperVersionFetcher(String githubBranchName, String downloadPage, String projectDisplayName, String organizationDisplayName, String gitHubOrganizationName, String gitHubRepoName) {
+        this.gitHubBranchName = githubBranchName;
+        this.downloadPage = downloadPage;
+        this.projectDisplayName = projectDisplayName;
+        this.organizationDisplayName = organizationDisplayName;
+        this.gitHubOrganizationName = gitHubOrganizationName;
+        this.gitHubRepoName = gitHubRepoName;
+    }
+
+    // Fiddle end - branding changes - version fetcher
+
+    @Override
+    public long getCacheTime() {
+        return 720000;
+    }
+
+    @Nonnull
+    @Override
+    public Component getVersionMessage(@Nonnull String serverVersion) {
+        // Fiddle start - branding changes - version fetcher
+        String[] parts = serverVersion.substring(("git-" + this.projectDisplayName + "-").length()).split("[-\\s]");
+        final Component updateMessage = getUpdateStatusMessage(this.gitHubOrganizationName + "/" + this.gitHubRepoName, this.gitHubBranchName, parts[0]);
+        // Fiddle end - branding changes - version fetcher
+        final Component history = getHistory();
+
+        return history != null ? TextComponent.ofChildren(updateMessage, Component.newline(), history) : updateMessage;
+    }
+
+    protected @Nullable String getMinecraftVersion() { // Fiddle - branding changes - version fetcher
+        if (mcVer == null) {
+            java.util.regex.Matcher matcher = VER_PATTERN.matcher(org.bukkit.Bukkit.getBukkitVersion());
+            if (matcher.find()) {
+                String result = matcher.group();
+                mcVer = result.substring(0, result.length() - 2); // strip 'R' anchor and trailing '-'
+            } else {
+                org.bukkit.Bukkit.getLogger().warning("Unable to match version to pattern! Report to " + this.organizationDisplayName + "!"); // Fiddle - branding changes - version fetcher
+                org.bukkit.Bukkit.getLogger().warning("Pattern: " + VER_PATTERN.toString());
+                org.bukkit.Bukkit.getLogger().warning("Version: " + org.bukkit.Bukkit.getBukkitVersion());
+            }
+        }
+
+        return mcVer;
+    }
+
+    // Fiddle start - branding changes - version fetcher
+
+    protected boolean canFetchDistanceFromSiteApi() {
+        return false;
+    }
+
+    protected int fetchDistanceFromSiteApi(int jenkinsBuild) {
+        return -1;
+    }
+
+    // Fiddle end - branding changes - version fetcher
+
+    private Component getUpdateStatusMessage(@Nonnull String repo, @Nonnull String branch, @Nonnull String versionInfo) {
+        // Fiddle start - branding changes - version fetcher
+        int distance = -1;
+        boolean readFromSiteApi = false;
+        if (this.canFetchDistanceFromSiteApi()) {
+        // Fiddle end - branding changes - version fetcher
+        try {
+            int jenkinsBuild = Integer.parseInt(versionInfo);
+            // Fiddle start - branding changes - version fetcher
+            distance = this.fetchDistanceFromSiteApi(jenkinsBuild);
+            readFromSiteApi = true;
+        } catch (NumberFormatException ignored) {}
+        }
+        if (!readFromSiteApi) {
+            // Fiddle end - branding changes - version fetcher
+            versionInfo = versionInfo.replace("\"", "");
+            distance = fetchDistanceFromGitHub(repo, branch, versionInfo);
+        }
+
+        switch (distance) {
+            case -1:
+                return Component.text("Error obtaining version information", NamedTextColor.YELLOW);
+            case 0:
+                return Component.text("You are running the latest version", NamedTextColor.GREEN);
+            case -2:
+                return Component.text("Unknown version", NamedTextColor.YELLOW);
+            default:
+                return Component.text("You are " + distance + " version(s) behind", NamedTextColor.YELLOW)
+                        .append(Component.newline())
+                        .append(Component.text("Download the new version at: ")
+                                .append(Component.text(this.downloadPage, NamedTextColor.GOLD) // Fiddle - branding changes - version fetcher
+                                        .hoverEvent(Component.text("Click to open", NamedTextColor.WHITE))
+                                        .clickEvent(ClickEvent.openUrl(this.downloadPage)))); // Fiddle - branding changes - version fetcher
+        }
+    }
+
+    // Contributed by Techcable <Techcable@outlook.com> in GH-65
+    private static int fetchDistanceFromGitHub(@Nonnull String repo, @Nonnull String branch, @Nonnull String hash) {
+        try {
+            HttpURLConnection connection = (HttpURLConnection) new URL("https://api.github.com/repos/" + repo + "/compare/" + branch + "..." + hash).openConnection();
+            connection.connect();
+            if (connection.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) return -2; // Unknown commit
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), Charsets.UTF_8))) {
+                JsonObject obj = new Gson().fromJson(reader, JsonObject.class);
+                String status = obj.get("status").getAsString();
+                switch (status) {
+                    case "identical":
+                        return 0;
+                    case "behind":
+                        return obj.get("behind_by").getAsInt();
+                    default:
+                        return -1;
+                }
+            } catch (JsonSyntaxException | NumberFormatException e) {
+                e.printStackTrace();
+                return -1;
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+            return -1;
+        }
+    }
+
+    @Nullable
+    private Component getHistory() {
+        final VersionHistoryManager.VersionData data = VersionHistoryManager.INSTANCE.getVersionData();
+        if (data == null) {
+            return null;
+        }
+
+        final String oldVersion = data.getOldVersion();
+        if (oldVersion == null) {
+            return null;
+        }
+
+        return Component.text("Previous version: " + oldVersion, NamedTextColor.GRAY, TextDecoration.ITALIC);
+    }
+}
diff --git a/src/main/java/org/fiddlemc/fiddle/version/FiddleVersionFetcher.java b/src/main/java/org/fiddlemc/fiddle/version/FiddleVersionFetcher.java
new file mode 100644
index 0000000000000000000000000000000000000000..714c2f94308f486f7e41d09ecb1849caa2901750
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/version/FiddleVersionFetcher.java
@@ -0,0 +1,11 @@
+// Fiddle - branding changes - version fetcher
+
+package org.fiddlemc.fiddle.version;
+
+public class FiddleVersionFetcher extends AbstractPaperVersionFetcher {
+
+    public FiddleVersionFetcher() {
+        super("ver/1.20.1", "https://github.com/FiddleMC/Fiddle", "Fiddle", "FiddleMC", "FiddleMC", "Fiddle");
+    }
+
+}
diff --git a/src/main/java/org/spigotmc/WatchdogThread.java b/src/main/java/org/spigotmc/WatchdogThread.java
index 98fba0288be9ed2cb18ffba5cf81148157dd4fcf..984096eb376ab62d7046f52835052b73e47b0c11 100644
--- a/src/main/java/org/spigotmc/WatchdogThread.java
+++ b/src/main/java/org/spigotmc/WatchdogThread.java
@@ -155,14 +155,20 @@ public final class WatchdogThread extends io.papermc.paper.util.TickThread // Pa
                 if (isLongTimeout) {
                 // Paper end
                 log.log( Level.SEVERE, "------------------------------" );
-                log.log( Level.SEVERE, "The server has stopped responding! This is (probably) not a Paper bug." ); // Paper
+                // Fiddle start - branding changes
+                /*
+                We do not want people to report thread issues to Paper,
+                but we do want people to report thread issues to Fiddle.
+                 */
+                log.log( Level.SEVERE, "The server has stopped responding! This is (probably) not a Paper bug. This could be a Fiddle bug." ); // Paper
+                // Fiddle end - branding changes
                 log.log( Level.SEVERE, "If you see a plugin in the Server thread dump below, then please report it to that author" );
                 log.log( Level.SEVERE, "\t *Especially* if it looks like HTTP or MySQL operations are occurring" );
                 log.log( Level.SEVERE, "If you see a world save or edit, then it means you did far more than your server can handle at once" );
                 log.log( Level.SEVERE, "\t If this is the case, consider increasing timeout-time in spigot.yml but note that this will replace the crash with LARGE lag spikes" );
-                log.log( Level.SEVERE, "If you are unsure or still think this is a Paper bug, please report this to https://github.com/PaperMC/Paper/issues" );
+                log.log( Level.SEVERE, "If you are unsure or think this is a Fiddle bug, please report this to https://github.com/FiddleMC/Fiddle/issues - and if you think this is a Paper bug, please report this to https://github.com/PaperMC/Paper/issues" ); // Fiddle - branding changes
                 log.log( Level.SEVERE, "Be sure to include ALL relevant console errors and Minecraft crash reports" );
-                log.log( Level.SEVERE, "Paper version: " + Bukkit.getServer().getVersion() );
+                log.log( Level.SEVERE, "Fiddle version: " + Bukkit.getServer().getVersion() ); // Fiddle - branding changes
                 //
                 if ( net.minecraft.world.level.Level.lastPhysicsProblem != null )
                 {
@@ -185,12 +191,18 @@ public final class WatchdogThread extends io.papermc.paper.util.TickThread // Pa
                 // Paper end
                 } else
                 {
-                    log.log(Level.SEVERE, "--- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH  - " + Bukkit.getServer().getVersion() + " ---");
+                    // Fiddle start - branding changes
+                    /*
+                    We do not want people to report thread issues to Paper,
+                    but we do want people to report thread issues to Fiddle.
+                     */
+                    log.log(Level.SEVERE, "--- DO NOT REPORT THIS TO PAPER - If you think this is a Fiddle bug, please report it at https://github.com/FiddleMC/Fiddle/issues - THIS IS NOT A PAPER BUG OR CRASH - " + Bukkit.getServer().getVersion() + " ---");
+                    // Fiddle end - branding changes
                     log.log(Level.SEVERE, "The server has not responded for " + (currentTime - lastTick) / 1000 + " seconds! Creating thread dump");
                 }
                 // Paper end - Different message for short timeout
                 log.log( Level.SEVERE, "------------------------------" );
-                log.log( Level.SEVERE, "Server thread dump (Look for plugins here before reporting to Paper!):" ); // Paper
+                log.log( Level.SEVERE, "Server thread dump (Look for plugins here before reporting to Fiddle!):" ); // Paper // Fiddle - branding changes
                 io.papermc.paper.chunk.system.scheduling.ChunkTaskScheduler.dumpAllChunkLoadInfo(isLongTimeout); // Paper // Paper - rewrite chunk system
                 this.dumpTickingInfo(); // Paper - log detailed tick information
                 WatchdogThread.dumpThread( ManagementFactory.getThreadMXBean().getThreadInfo( MinecraftServer.getServer().serverThread.getId(), Integer.MAX_VALUE ), log );
@@ -206,7 +218,13 @@ public final class WatchdogThread extends io.papermc.paper.util.TickThread // Pa
                     WatchdogThread.dumpThread( thread, log );
                 }
                 } else {
-                    log.log(Level.SEVERE, "--- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH ---");
+                    // Fiddle start - branding changes
+                    /*
+                    We do not want people to report thread issues to Paper,
+                    but we do want people to report thread issues to Fiddle.
+                     */
+                    log.log(Level.SEVERE, "--- DO NOT REPORT THIS TO PAPER - If you think this is a Fiddle bug, please report it at https://github.com/FiddleMC/Fiddle/issues - THIS IS NOT A PAPER BUG OR CRASH ---");
+                    // Fiddle end - branding changes
                 }
 
                 log.log( Level.SEVERE, "------------------------------" );
